extensions/src/data_bridge/index.ts

```diff
@@ -15,77 +19,99 @@ const plugin: JupyterFrontEndPlugin<void> = {
-            window.parent.postMessage(
-                {
-                    type: "from-iframe-to-host",
-                    requestData: true,
-                    variableName,
-                },
-                "*"
-            );
+        window.sendDataToHost = (data: object) => {
+            const MESSAGE_SET_DATA_CONTENT = {
+                type: "from-iframe-to-host",
+                action: "set-data",
+                payload: data
+            };
+            window.parent.postMessage(MESSAGE_SET_DATA_CONTENT, "*");
         };
 
-        window.addEventListener("message", async (event) => {
-            if (event.data.type === "from-host-to-iframe") {
-                let data = event.data.data;
-                let variableName = event.data.variableName || "data";
-                const dataJson = JSON.stringify(data);
-                const code = `
-  import json
-  ${variableName} = json.loads('${dataJson}')
-      `;
-                // Similar to https://jupyterlab.readthedocs.io/en/stable/api/classes/application.LabShell.html#currentWidget
-                // https://jupyterlite.readthedocs.io/en/latest/reference/api/ts/interfaces/jupyterlite_application.ISingleWidgetShell.html#currentwidget
-                const currentWidget = app.shell.currentWidget;
-
-                if (currentWidget instanceof NotebookPanel) {
-                    const notebookPanel = currentWidget;
-                    const kernel = notebookPanel.sessionContext.session?.kernel;
+        /**
+         * Listen for messages from the host page, and update the data in the kernel
+         * @param event MessageEvent
+         */
+        window.addEventListener(
+            "message",
+            async (event: MessageEvent<IframeMessageSchema>) => {
+                if (event.data.type === "from-host-to-iframe") {
+                    dataFromHost = JSON.stringify(event.data.payload);
+                    const notebookPanel = notebookTracker.currentWidget;
+                    await notebookPanel?.sessionContext.ready;
+                    const sessionContext = notebookPanel?.sessionContext;
+                    const kernel = sessionContext?.session?.kernel;
                     if (kernel) {
-                        kernel.requestExecute({ code: code });
-                    } else {
-                        console.error("No active kernel found");
+                        loadData(kernel, dataFromHost);
                     }
-                } else {
-                    console.error("Current active widget is not a notebook");
                 }
             }
-        });
-    },
+        );
+
+        /**
+         * Load the data into the kernel by executing code
+         * @param kernel
+         * @param data string representation of JSON
+         */
+        const loadData = (kernel: IKernelConnection, data: string) => {
+            const code = `import json\ndata_from_host = json.loads("${data}")`;
```

Comment:
Let's use the approach from https://github.com/Exabyte-io/standata/blob/main/build_runtime_data.js#L67C8-L67C91